File: 2024/02/decode-string/1.py
# https://leetcode.cn/problems/decode-string/

class Solution:
    def decodeString(self, s: str) -> str:
        if s.count(']') == 0:
            return s
        # 找到第一个 ]
        right = s.find(']')
        # 找到前面的 [
        left = s.rfind('[',0,right)
        # 匹配前面的数字
        num = ''
        idx=left
        while True:
            idx-=1
            if s[idx].isdigit() is False:
                break
            num = s[idx] + num
        # 展开
        flatten = s[left+1:right]*int(num)
        res = s[:idx+1] + flatten + s[right+1:]
        # 递归
        print(flatten, res)
        return self.decodeString(res)

import re
class Solution:
    def decodeString(self, s: str) -> str:
        if s.find(']') == -1:
            return s
        match = re.search(r'(\d+)\[(\w+)\]', s)
        res = s.replace(match[0], match[2] * int(match[1]))
        return self.decodeString(res)

File: 2024/02/binary-search/1.js
// https://leetcode.cn/problems/binary-search/

/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var search = function (nums, target) {
    let left = 0, right = nums.length - 1;
    while (right >= left) {
        let mid = ~~((left + right) / 2);
        if (nums[mid] === target)
            return mid;
        if (nums[mid] < target)
            left = mid + 1;
        if (nums[mid] > target)
            right = mid - 1;
    }
    return -1;
};

// nums = [-1, 0, 3, 5, 9, 12], target = 9
// nums = [5], target = 5
// nums = [-1, 0, 3, 5, 9, 12], target = 2
// console.log(search(nums, target));